### PR TITLE
Add first module (crypto) and project plumbing

### PR DESCRIPTION
**Pull request by kgriffs**

### CODE REVIEW

**Comment by reaperhulk** [comment]
The way the travis.yml is currently configured you're actually running both the 2.7 and 3.4 tox jobs in both the 2.7 and 3.4 travis jobs. tox will automatically pick up $TOXENV (if present) and pass it as if you used `-e` but the python environments in travis don't do that. I'd recommend expanding the Python section to look like this:

```
python:
  - '2.7'
    env: TOXENV=py27
  - '3.4'
    env: TOXENV=py34
```

File: rmsutil/crypto.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import os
+
+import six
+
+
+def get_rand_hex(entropy_bytes):
+    """Returns a random hexadecimal string.
+
+    Args:
+        entropy_bytes (str): Number of bytes of entropy to encode.
```
Comment (reaperhulk) [inline rmsutil/crypto.py]: The arg here is not a string

File: rmsutil/crypto.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    Args:
+        key (str): 128-bit signing key encoded in hexadecimal.
+        message (str or bytes): Arbitrary message to hash. If a Unicode
+            string, will be encoded as UTF-8 before hashing.
+
+    Returns:
+        str: Hexadecimal-encoded HMAC of the message
+    """
+
+    if len(key) != 32:  # 128 / 8 * 2
```
Comment (reaperhulk) [inline rmsutil/crypto.py]: Why does this require it to be hex encoded? Also, why does it require the key be 128-bit? Presumably to avoid the hashing step if the key is too long, but the "block" size of SHA256 is 512-bit so this restriction makes little sense. Maybe the restriction should be not less than 128-bit?

**Comment by kgriffs** [comment]
@reaperhulk I've seen the below as well. I guess I better go read up on the travis docs...

``` yaml
env:
  - TOXENV=py27
  - TOXENV=py34
```

File: rmsutil/crypto.py (hunk 1 above)
Comment (kgriffs) [inline rmsutil/crypto.py]: Fixed.

File: rmsutil/crypto.py (hunk 2 above)
Comment (kgriffs) [inline rmsutil/crypto.py]: > Maybe the restriction should be not less than 128-bit?

Good suggestion; that makes sense.

> Why does this require it to be hex encoded?

Alternatively, we could use base64. Regardless, I didn't think it beneficial to require raw bytes, since it's more convenient for apps to just use encoded keys without having to decode them each time they are used.

**Comment by kgriffs** [comment]
OK, so it looks like if you leave off "python" it will default to 2.7. And then you can specify a list of TOXENV in env to cause travis to perform one tox run for py27 and py34.

**Comment by reaperhulk** [comment]
Yep, if you don't specify a python the travis UI will always say "2.7" though :) I made a mistake in my previous in leaving it as a python section, but I'd recommend:

```
env:
  - python: 2.7
    TOXENV=py27
```

It's what we do in cryptography (https://github.com/pyca/cryptography/blob/master/.travis.yml#L13). Nevermind the matrix include thing...that's because we have way more complex requirements than normal :grinning:

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+sudo: false
+language: python
+env:
+  - TOX_ENV=py27
```
Comment (reaperhulk) [inline .travis.yml]: The variable needs to be TOXENV. TOX_ENV is not picked up by tox so all 3 targets are still being run :)

Comment (kgriffs) [inline .travis.yml]: Fixed.

**Comment by reaperhulk** [comment]
Okay you've addressed my comments. I think there's a more meta-comment about what we want out of functions like this though. Is it really a commonly reusable pattern to want to pass a hex key for doing an HMAC?

I'd definitely like to see others weigh in on this PR before we merge. :)

**Comment by kgriffs** [comment]
Suggestions are welcome. This came out of my need to share code between racksack and olivaw to HMAC an account ID. I wanted to have an opaque string to use in the path for downloading customer-specific falcon sensor packages. With this approach, the hex key would be provided via an env variable in the case of olivaw, and entered via getpass in the case of racksack.

**Comment by kgriffs** [comment]
Options:
- Hex (as-is). Caller doesn't have to decode to bytes before calling the function.
- URL-safe Base64 a la Fernet. Similar to above, but results in shorter key strings.
- Bytes. Most flexible (app can decide how they want to encode their keys), but also gives at least some of our apps more work to do, since they will be configured with textual keys.

**Comment by kgriffs** [comment]
Removed rand_hex; I pulled it out of racksack but uuid4 is a simpler option for how that function was being used.

**Comment by lvh** [comment]
Have a bytes low-level API, mark as low-level; encode as URL-safe base64 by default.

**Comment by kgriffs** [comment]
Something like this?

``` py
rmsutil.crypto.hmac_sha256(key_as_base64, message_as_str_or_bytes)
rmsutil.crypto.substrate.hmac_sha256(key_as_bytes, message_as_bytes)
```

**Comment by reaperhulk** [comment]
Why not just require bytes? I think fernet's base64 encoding is a mistake honestly. Added complexity for extremely limited benefit (and in many cases it's actively a disadvantage since it bloats the payload even when you don't need it).

**Comment by lvh** [comment]
Well, you have the option of using bytes. My reasoning for using URLsafe b64 in e.g. txscrypt is that you can dump it _nearly_ anywhere without worrying about it too much. While people _should_ probably know about which formats they're using and if it's okay to store arbitrary bytes in there or not, that doesn't mean they do, and that's particularly shitty when the problem only occurs _sometimes_ (like when your output contains a nul byte, or a comma, or...).

I think issues with finding appropriate encodings are very common. Consider JSON. All characters that base64 produces are encodable in a single byte. Because JSON strings are unicode, not bytes, nul bytes (and many others) are encoded as:

```
"\u0000"
```

That's 6 bytes to encode 1 byte. The number of directly printable characters is:

```
In [8]: len(string.ascii_letters + string.digits + string.punctuation)
Out[8]: 94
```

That's only 94/256, or about 1/3rd. Some of those need to be encoded as multiple bytes, like solidus, reverse solidus. Some of those are directly encodable, like backspace et cetera. If we split the difference, that results in an approximate average bytes of storage per byte encoded of:

(94/256 \* 1) + ((256 - 94)/256 \* 6) ~= 4.16

Which is much higher than base63's approximate 3/2, all of which are nice and common characters that are likely to compress back down quite well with plain old Huffman coding, which is most compression formats you're likely to see for both storage and transport.

However, the behavior than John Q User who just wants to see this work is more likely to see is that, high bytes will just _blow up_:

```
>>>> json.dumps("\x89")
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-9-495c4d6913bf> in <module>()
----> 1 json.dumps("\x89")

/Users/lvh/.pyenv/versions/2.7.10/lib/python2.7/json/__init__.pyc in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, encoding, default, sort_keys, **kw)
    241         cls is None and indent is None and separators is None and
    242         encoding == 'utf-8' and default is None and not sort_keys and not kw):
--> 243         return _default_encoder.encode(obj)
    244     if cls is None:
    245         cls = JSONEncoder

/Users/lvh/.pyenv/versions/2.7.10/lib/python2.7/json/encoder.pyc in encode(self, o)
    199                     o = o.decode(_encoding)
    200             if self.ensure_ascii:
--> 201                 return encode_basestring_ascii(o)
    202             else:
    203                 return encode_basestring(o)

UnicodeDecodeError: 'utf8' codec can't decode byte 0x89 in position 0: invalid start byte
```

So, in conclusion, while I agree that a low-level bytes API being available is a good thing for when you're absolutely sure, I think the no-sharp-edges default should be urlsafe base64.

**Comment by kgriffs** [comment]
How generic do we want or need to make this function? I would argue that we already have a generic, low-level API in the form of the standard lib hmac and hashlib modules (or, alternatively, in cryptography.hazmat).

No-sharp-edges argument aside, for the use cases I've got in mind, requiring bytes would slightly reduce the utility of the function, since we'll likely just configure the apps with encoded keys for convenience. The "message" will always be a small bit of text, such as a customer ID. Whenever we need to do anything more complicated, we'll instead use a well-vetted recipe such as GPG (PGP) or Fernet.

P.S. - Perhaps the function should be renamed to reflect the narrow scope of its anticipated use...

**Comment by reaperhulk** [comment]
Renaming the function :+1: 

On the base64 topic, we're afield of what matters for this PR, but I'll note that Python's policy of being extremely liberal in what it accepts for decoding can cause [confusion](https://github.com/pyca/cryptography/issues/1994). :)

File: rmsutil/crypto.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    """Calculate an opaque ID corresponding to a private identifier.
+
+    This function uses HMAC-SHA256 to hash a private identifier to an
+    unpredictable, opaque ID. The hashed ID can be used in place of the
+    original in order to avoid exposing a potentially sensitive value,
+    such as a database table's monotonic primary key, in an insecure
+    context.
+
+    Args:
+        key (six.text_type): URL-safe Base64-encoded HMAC key, at least
```
Comment (kgriffs) [inline rmsutil/crypto.py]: I'm thinking that for RMS projects that need to be compatible with py27 (not just py3k), we should standardize on `unicode` strings.

Comment (reaperhulk) [inline rmsutil/crypto.py]: I am neutral on that as a requirement. It is consistent, but can be confusing since `""` is implicitly `u""` on py3.

File: rmsutil/crypto.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            16 bytes in length.
```
Comment (lvh) [inline rmsutil/crypto.py]: I'm guessing this is supposed to be 16 bytes _before b64ing_, which isn't how this reads to me as currently worded.

Comment (kgriffs) [inline rmsutil/crypto.py]: That's correct; I'll clarify.

Comment (lvh) [inline rmsutil/crypto.py]: I would say "128 bits of entropy" but that may not be particularly enlightening to the end-user :)

Comment (kgriffs) [inline rmsutil/crypto.py]: Updated the comment.

**Comment by lvh** [comment]
LGTM!
